### PR TITLE
(3) Signing Refactor - Refactor crypto tests and make recoverAddress async

### PR DESCRIPTION
**Pull request by pedrouid**
requires #910

### CODE REVIEW

File: modules/cf-core/src/machine/middleware.ts [hunk 1]
```diff
@@ -18,8 +18,8 @@ export class MiddlewareContainer {
   public async run(opCode: Opcode, args: any[]) {
     const middleware = this.middlewares[opCode][0];
 
-    if (middleware === undefined) {
-      throw Error(`Attempted to run middleware for opcode ${opCode} but none existed`);
+    if (typeof middleware === "undefined") {
+      throw new Error(`Attempted to run middleware for opcode ${opCode} but none existed`);
```
Comment (LayneHaber) [inline modules/cf-core/src/machine/middleware.ts]: this won't catch nulls

File: modules/cf-core/src/machine/protocol-runner.ts [hunk 2]
```diff
@@ -76,8 +76,8 @@ export class ProtocolRunner {
   public async runProtocolWithMessage(msg: ProtocolMessage) {
     const protocol = getProtocolFromName(msg.protocol);
     const step = protocol[msg.seq];
-    if (step === undefined) {
-      throw Error(`Received invalid seq ${msg.seq} for protocol ${msg.protocol}`);
+    if (typeof step === "undefined") {
+      throw new Error(`Received invalid seq ${msg.seq} for protocol ${msg.protocol}`);
```
Comment (LayneHaber) [inline modules/cf-core/src/machine/protocol-runner.ts]: same as above

File: modules/cf-core/src/methods/app-instance/take-action/controller.ts [hunk 3]
```diff
@@ -38,7 +38,7 @@ export default class TakeActionController extends NodeController {
     const { appInstanceId, action } = params;
 
     if (!appInstanceId) {
-      throw Error(NO_APP_INSTANCE_FOR_TAKE_ACTION);
+      throw new Error(NO_APP_INSTANCE_FOR_TAKE_ACTION);
```
Comment (LayneHaber) [inline modules/cf-core/src/methods/app-instance/take-action/controller.ts]: this looks so tedious im sorry lol

File: modules/cf-core/src/machine/middleware.ts (hunk 1 above)
Comment (pedrouid) [reply modules/cf-core/src/machine/middleware.ts]: but the previous one wouldnt either
```javascript
const testNull = null

null === undefined
> false

typeof testNull === 'undefined'
> false
```

File: modules/cf-core/src/methods/app-instance/take-action/controller.ts (hunk 3 above)
Comment (pedrouid) [reply modules/cf-core/src/methods/app-instance/take-action/controller.ts]: nah it was just global replace

File: modules/cf-core/src/machine/protocol-runner.ts (hunk 2 above)
Comment (pedrouid) [reply modules/cf-core/src/machine/protocol-runner.ts]: read comment above

**Pull request merged**